sigpro/pipeline.py

```diff
@@ -0,0 +1,596 @@
+                        output_column_name += '.' + str(layer)
+                        for output_dict in final_primitive.get_outputs():
+                            out_name = output_dict['name']
+                            final_primitive_outputs[numbered_primitive_name][out_name] = \
+                                f'{output_column_name}.' + str(out_name)
+                    else:
+                        npn = numbered_primitive_name[:]  # lint
+                        for output_dict in final_primitive.get_outputs():
+                            out_name = output_dict['name']
+                            final_outputs.append({'name': output_column_name + '.' + str(out_name),
+                                                  'variable': f'{npn}.{out_name}'})
+        return MLPipeline(
+            primitives=final_primitives_list,
+            init_params=final_init_params,
+            input_names=final_primitive_inputs,
+            output_names=final_primitive_outputs,
+            outputs={'default': final_outputs}
+        )
+
+    def get_primitives(self):
+        """Get a list of primitives in the pipeline."""
+        return self.primitives.copy()
+
+    def get_output_combinations(self):
+        """Get a list of output feature tuples produced by the pipeline."""
+        return [x[:] for x in self.primitive_combinations]
+
+
+def build_tree_pipeline(transformation_layers, aggregation_layer):
+    """
+    Build a tree pipeline using given transformation and aggregation layers.
+
+    Args:
+        transformation_layers (list): List of transformation layers, each
+        a list of transformation primitives.
+
+        aggregation_layer (list): List of aggregation primitives
+
+    Returns:
+        sigpro.pipeline.LayerPipeline:
+            A ``LayerPipeline`` object that produces the features in the Cartesian product
+            of all transformation layers and the aggregation layer.
+    """
+    primitives_all = set()
+    all_layers = []
+
+    if not isinstance(transformation_layers, list):
+        raise ValueError('transformation_layers must be a list')
+    for layer in transformation_layers:
+        if isinstance(layer, list):
+            for primitive_ in layer:
+                if not isinstance(primitive_, Primitive):
+                    raise ValueError('Non-primitive specified in transformation_layers')
+            all_layers.append(layer.copy())
+            primitives_all.update(layer)
+        else:
+            raise ValueError('Each layer in transformation_layers must be a list')
+    if isinstance(aggregation_layer, list):
+        for primitive_ in aggregation_layer:
+            if not isinstance(primitive_, Primitive):
+                raise ValueError('Non-primitive specified in aggregation_layer')
+        all_layers.append(aggregation_layer.copy())
+        primitives_all.update(aggregation_layer)
+    else:
+        raise ValueError('aggregation_layer must be a list')
```

Comment:
nit: to increase readability, add an empty line after any indentation reduction. There is a special case for if statements but in general, this is the style to follow:
```python3
if condition:
    raise an error
else: # notice here there is no empty line because the if statement is contained
    continue

if condition:
    ans = []
    for value in array:
        x = f(value)
        ans.append(x)

    result = sum(ans) # added empty line

else: # added another empty line
    continue
```

Reply:
Readability changes have been added to pipeline.py